lgc/builder/BuilderImpl.cpp

```diff
@@ -323,6 +323,150 @@ BranchInst *BuilderImplBase::createIf(Value *condition, bool wantElse, const Twi
   return branch;
 }
 
+#if defined(LLVM_HAVE_BRANCH_AMD_GFX)
+// =====================================================================================================================
+// For a non-uniform input, try and trace back through a descriptor load to
+// find the non-uniform index used in it. If that fails, we just use the
+// operand value as the index.
+//
+// Note: this code has to cope with relocs as well, this is why we have to
+// have a worklist of instructions to trace back
+// through. Something like this:
+// %1 = call .... @lgc.descriptor.set(...)          ;; Known uniform base
+// %2 = call .... @llvm.amdgcn.reloc.constant(...)  ;; Known uniform reloc constant
+// %3 = ptrtoint ... %1 to i64
+// %4 = zext ... %2 to i64
+// %5 = add i64 %3, %4
+// %6 = inttoptr i64 %5 to ....
+// %7 = bitcast .... %6 to ....
+// %8 = getelementptr .... %7, i64 %offset
+//
+// As long as the base pointer %7 can be traced back to a descriptor set and
+// reloc we can infer that it is truly uniform and use the gep index as the waterfall index safely.
+//
+// @param nonUniformVal : Value representing non-uniform descriptor
+// @return : Value representing the non-uniform index
+static Value *traceNonUniformIndex(Value *nonUniformVal) {
+  auto load = dyn_cast<LoadInst>(nonUniformVal);
+  if (!load)
+    return nonUniformVal;
+
+  SmallVector<Value *, 2> worklist;
+  Value *base = load->getOperand(0);
+  Value *index = nullptr;
+
+  // Loop until a descriptor table reference or unexpected operation is reached.
+  // In the worst case this may visit all instructions in a function.
+  for (;;) {
+    if (auto bitcast = dyn_cast<BitCastInst>(base)) {
+      base = bitcast->getOperand(0);
+      continue;
+    }
+    if (auto gep = dyn_cast<GetElementPtrInst>(base)) {
+      if (gep->hasAllConstantIndices()) {
+        base = gep->getPointerOperand();
+        continue;
+      }
+      // Variable GEP, to provide the index for the waterfall.
+      if (index || gep->getNumIndices() != 1)
+        break;
+      index = *gep->idx_begin();
+      base = gep->getPointerOperand();
+      continue;
+    }
+    if (auto extract = dyn_cast<ExtractValueInst>(base)) {
+      if (extract->getIndices().size() == 1 && extract->getIndices()[0] == 0) {
+        base = extract->getAggregateOperand();
+        continue;
+      }
+      break;
+    }
+    if (auto insert = dyn_cast<InsertValueInst>(base)) {
+      if (insert->getIndices()[0] != 0) {
+        base = insert->getAggregateOperand();
+        continue;
+      }
+      if (insert->getIndices().size() == 1 && insert->getIndices()[0] == 0) {
+        base = insert->getInsertedValueOperand();
+        continue;
+      }
+      break;
```

Comment:
Could we add test(s) that exercise these two if that's something quick to write?

Reply:
Ping

Reply:
I spent a while looking at this.
I found it is impossible to trigger these branches at the moment because any relevant insert and extract operations are simplified out by an instcombine pass executed before the image builder (waterfall creation).

I am going to guess they were written before the instcombine pass was present.
Just to be sure I ran a library of existing pipelines against the compiler and these branches were never triggered.
So at the moment it seems impossible to exercise these via a test.

Reply:
Thanks for looking into this